prometheus/collectors/dbstats_collector.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build go1.15
+// +build go1.15
+
+package collectors
+
+import (
+	"database/sql"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type dbStatsCollector struct {
+	db *sql.DB
+
+	maxOpenConnections *prometheus.Desc
+
+	openConnections  *prometheus.Desc
+	inUseConnections *prometheus.Desc
+	idleConnections  *prometheus.Desc
+
+	waitCount         *prometheus.Desc
+	waitDuration      *prometheus.Desc
+	maxIdleClosed     *prometheus.Desc
+	maxIdleTimeClosed *prometheus.Desc
+	maxLifetimeClosed *prometheus.Desc
+}
+
+// DBStatsCollectorOpts defines the behavior of a db stats collector
+// created with NewDBStatsCollector.
+type DBStatsCollectorOpts struct {
+	// DriverName holds the name of driver.
+	// It will not used for empty strings.
+	DriverName string
+}
+
+// NewDBStatsCollector returns a collector that exports metrics about the given *sql.DB.
+// See https://golang.org/pkg/database/sql/#DBStats for more information on stats.
+func NewDBStatsCollector(db *sql.DB, opts DBStatsCollectorOpts) prometheus.Collector {
+	var fqName func(name string) string
+	if opts.DriverName == "" {
+		fqName = func(name string) string {
+			return "go_db_stats_" + name
+		}
+	} else {
+		fqName = func(name string) string {
+			return "go_" + opts.DriverName + "_db_stats_" + name
```

Comment:
I'm pretty sure the driver name should not be part of the metric name. It should be a label, perhaps called `db_name`.

That label has to be a const label when creating the descriptors, see comment below.